src/Streamly/Internal/Data/Stream/Async/Channel/Dispatcher.hs

```diff
@@ -243,85 +280,93 @@ dispatchWorkerPaced sv = do
-        -- deadlock. Deadlocks may still happen if for some weird reason
-        -- the consuming computation shares an MVar or some other resource
-        -- with the producing computation and gets blocked on that resource
-        -- and therefore cannot do any pushworker to add more threads to
-        -- the producer. In such cases the programmer should use a parallel
-        -- style so that all the producers are scheduled immediately and
-        -- unconditionally. We can also use a separate monitor thread to
-        -- push workers instead of pushing them from the consumer, but then
-        -- we are no longer using pull based concurrency rate adaptation.
-        --
-        -- XXX update this in the tutorial.
-        --
-        -- Having pending active workers does not mean that we are guaranteed
-        -- to be woken up if we sleep. In case of Ahead streams, there may be
-        -- queued items in the heap even though the outputQueue is empty, and
-        -- we may have active workers which are deadlocked on those items to be
-        -- processed by the consumer. We should either guarantee that any
-        -- worker, before returning, clears the heap or we send a worker to
-        -- clear it. Normally we always send a worker if no output is seen, but
-        -- if the thread limit is reached or we are using pacing then we may
-        -- not send a worker. See the concurrentApplication test in the tests,
-        -- that test case requires at least one yield from the producer to not
-        -- deadlock, if the last workers output is stuck in the heap then this
-        -- test fails.  This problem can be extended to n threads when the
-        -- consumer may depend on the evaluation of next n items in the
-        -- producer stream.
-
-        -- register for the outputDoorBell before we check the queue so that if
-        -- we sleep because the queue was empty we are guaranteed to get a
-        -- doorbell on the next enqueue.
-
-        liftIO $ atomicModifyIORefCAS_ (needDoorBell sv) $ const True
-        liftIO storeLoadBarrier
-        canDoMore <- dispatch sv
-
-        -- XXX test for the case when we miss sending a worker when the worker
-        -- count is more than 1500.
-        --
-        -- XXX Assert here that if the heap is not empty then there is at
-        -- least one outstanding worker. Otherwise we could be sleeping
-        -- forever.
-
-        if canDoMore
-        then sendWorkerWait delay dispatch sv
-        else do
-            liftIO
-                $ withDiagMVar
-                    (svarInspectMode sv)
-                    (dumpSVar sv)
-                    "sendWorkerWait: nothing to do"
-                $ takeMVar (outputDoorBell sv)
-            (_, len) <- liftIO $ readIORef (outputQueue sv)
-            when (len <= 0) $ sendWorkerWait delay dispatch sv
+sendWorkerWait eager delay dispatch sv = go
+
+    where
+
+    go = do
```

Comment:
What was the point of this change? 
You've essentially pushed the function into a go-loop.
Any specific reason?

Reply:
this is static argument transformation, the commit message states the reason of the change. It is performance.

Reply:
sendWorkerWait was a recursive function threading around all its arguments in the recursive calls. With this change no arguments are passed in the recursive call.